### PR TITLE
[ETCM-709] Improve ommers validations.

### PR DESCRIPTION
**Pull request by leo-bogastry**
# Description
In testnet we observed the following error message `Block import error BlockImportFailed(OmmersAncestorsError)`, but looking at the code it was not possible to know exactly why the validation failed.
The goal of this PR is to allow for better debugging in case we observe the same error message again.

# Proposed Solution

Replaced OmmersAncestorsError by more specific OmmerIsAncestorError and OmmerParentIsNotAncestorError.
Renamed OmmersNotValidError to OmmersHeaderError and added the underlying reason for failure 
Refactored validateOmmersAncestors method.
Refactored validateOmmersHeaders method to not hide the underlying failing reason.
Reorganized some packages because some classes were not in the correct package and/or the Spec was not in the correct package. Renamed some specs to better reflect the class being tested
Raised the log level of some logs in BlockFetcher because there are barely any logs this actor when debug logs are disabled

# Testing
Connected to Sagano with FastSync off and RegularSync is importing blocks properly

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 1]
```diff
@@ -144,19 +152,17 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
   private def collectOmmersFromAncestors(
       parentHash: ByteString,
       blockNumber: BigInt,
       getNBlocksBack: GetNBlocksBack
-  ): Option[Seq[BlockHeader]] = {
+  ): Seq[BlockHeader] = {
     val numberOfBlocks = blockNumber.min(OmmerGenerationLimit).toInt
-    val ancestors = getNBlocksBack(parentHash, numberOfBlocks).map(_.body.uncleNodesList)
-    Some(ancestors).filter(_.length == numberOfBlocks).map(_.flatten)
+    getNBlocksBack(parentHash, numberOfBlocks).flatMap(_.body.uncleNodesList)
```
Comment (leo-bogastry) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I have simplified both `collectAncestors` and `collectOmmersFromAncestors` methods, by removing the final check.
It was checking that the `getNBlocksBack` did return the expected number of items, otherwise it would just return None.
Is there any reason to keep such a check?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala [hunk 2]
```diff
@@ -113,9 +114,10 @@ class BlockFetcher(
       }
 
       fetchBlocks(newState)
+
     case InvalidateBlocksFrom(blockNr, reason, withBlacklist) =>
       val (blockProvider, newState) = state.invalidateBlocksFrom(blockNr, withBlacklist)
-      log.debug("Invalidate blocks from {}", blockNr)
+      log.info("Invalidate blocks from {}", blockNr)
```
Comment (dzajkowski) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: I still think that these should not be changed to info. a better approach (imo) would be to setup the testnet with the logbacks BlockFetcher entry set to debug. this way we avoid burdening every other usage of mantis with additional logging.

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: That is indeed a valid approach. I do find weird that BlockFetcher has so many debug log entries, while the rest of the application also uses info. Note: I did not change all logs in BlockFetcher to be info

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala [hunk 3]
```diff
@@ -149,17 +151,17 @@ class BlockFetcher(
...
+      log.info("Something failed on a headers request, cancelling the request and re-fetching")
 
       val newState = state.withHeaderFetchReceived
       fetchBlocks(newState)
   }
 
   private def handleBodiesMessages(state: BlockFetcherState): Receive = {
     case Response(peer, BlockBodies(bodies)) if state.isFetchingBodies =>
-      log.debug(s"Received ${bodies.size} block bodies")
+      log.debug("Received {} block bodies", bodies.size)
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala:162]: So this is a situation where this change improves performance, due to actor logging not being based on macro's, correct?

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 4]
```diff
@@ -66,16 +66,22 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
   ): Either[OmmersError, OmmersValid] = {
-    if (ommers.forall(blockHeaderValidator.validate(_, getBlockHeaderByHash).isRight)) Right(OmmersValid)
-    else Left(OmmersNotValidError)
+    val validationsResult = ommers.map(blockHeaderValidator.validate(_, getBlockParentsHeaderByHash))
+
+    if (validationsResult.forall(_.isRight)) Right(OmmersValid)
+    else {
+      validationsResult.filterNot(_.isRight) match {
+        case Left(headerError) :: _ => Left(OmmersHeaderError(Some(headerError)))
+        case _ => Left(OmmersHeaderError(None))
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: What does this case mean? Is there any way to make this more informative?

File: src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala [hunk 5]
```diff
@@ -36,9 +37,10 @@ object OmmersValidator {
 
   object OmmersError {
     case object OmmersLengthError extends OmmersError
-    case object OmmersNotValidError extends OmmersError
+    case class OmmersHeaderError(error: Option[BlockHeaderError]) extends OmmersError
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala]: An optional argument to an error class seems a bit weird to me as it's so uninformative. (Also commented below about this.)

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 6]
```diff
@@ -85,26 +91,28 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
+  private[std] def validateOmmersAncestors(
       parentHash: ByteString,
       blockNumber: BigInt,
       ommers: Seq[BlockHeader],
       getNBlocksBack: GetNBlocksBack
   ): Either[OmmersError, OmmersValid] = {
 
-    val ancestorsOpt = collectAncestors(parentHash, blockNumber, getNBlocksBack)
+    val ancestors = collectAncestors(parentHash, blockNumber, getNBlocksBack)
+    def ommersHashes: Seq[ByteString] = ommers.map(_.hash)
```
Comment (jvdp) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: Would it make sense to use `val` instead of `def`, and `lazy val` for some bindings that are not always used?

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: Ideally this would've been covered by tests if it was important. I can't comment on the check itself.

What leads you to believe it's not necessary to keep? Perhaps we can leave an assert in place, just in case?

File: src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala [hunk 7]
```diff
@@ -77,7 +92,29 @@ class OmmersValidatorSpec extends AnyFlatSpec with Matchers with ScalaCheckPrope
   // scalastyle:off magic.number
   trait BlockUtils extends EphemBlockchainTestSetup {
 
-    val ommersValidator = new StdOmmersValidator(blockchainConfig, new EthashBlockHeaderValidator(blockchainConfig))
+    val ommersValidator = new StdOmmersValidator(new EthashBlockHeaderValidator(blockchainConfig))
+
+    val ommerInvalidBranch = BlockHeader(
+      parentHash = ByteString(Hex.decode("fd07e36cfaf327801e5696134b12345f6a89fb1e8f017f2411a29d0ae810ab8b")),
```
Comment (jvdp) [inline src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala:112]: Why did all this have to be added? What do all the scary hex strings mean?

**Review by jvdp — APPROVED** [review]
Better errors and simpler code is better 👍 

I have no opinion about the log level stuff. I think we might want to bump default log level for end users to warning anyway? There's room for a story to really think about what we want out of logging, I think.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I cannot find a single reason to keep it, it adds noise to the code. The method was returning an `Option[List]` which is quite redundant and makes everything more hard to read

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala (hunk 3 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala:162]: I had some colleagues that had a very strong opinion on using templates, but I researched about this last Friday and apparently does not make a difference anymore :/

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 4 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: I had something more specific here, but then I got the build to fail because it was matching this to an error

File: src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala (hunk 5 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala]: The internal BlockHeaderError was being completely discarded and replaced by a generic `OmmersNotValidError`. This error happens when during the ommers header validation and it can fail because of 18 different reasons, and that will be the specific BlockHeaderError.
It's an `Option` for code reasons, programmatically has to cover all possible cases.
This may indeed a code smell, why is there a test that returns a generic error running this code? I did not check to be honest.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 6 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: `lazy val` can also be. Those pieces of code only run 0 or 1 time, def is just a more compact notation :)

File: src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala (hunk 7 above)
Comment (leo-bogastry) [reply src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala:112]: It's for test I added. The scary hex'es are hashes :)

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 6 above)
Comment (jvdp) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: A future refactoring could inadvertently cause some of this to run 1+ times, though.

File: src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala (hunk 7 above)
Comment (jvdp) [reply src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala:112]: Could these hashes be computed on the fly so you can see what they're based on? Are these from existing tests?

Comment (leo-bogastry) [reply src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala:112]: These are not random hashes, they validate the block. If you notice I had to write a different test, than the rest of the tests in the spec, trying to just make one hash invalid was breaking the whole block and the validation was failing before validating the ommers.
I did copy this data from another test, yes, and it made me realize that getting this kind of data is not trivial. Maybe it's available in block explorers, I did not check, it just crossed my mind

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 6 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: I have updated it to `lazy val`. Future proof :)

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 8]
```diff
@@ -66,16 +66,22 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
+      getBlockParentsHeaderByHash: GetBlockHeaderByHash
   ): Either[OmmersError, OmmersValid] = {
-    if (ommers.forall(blockHeaderValidator.validate(_, getBlockHeaderByHash).isRight)) Right(OmmersValid)
-    else Left(OmmersNotValidError)
+    val validationsResult = ommers.map(blockHeaderValidator.validate(_, getBlockParentsHeaderByHash))
+
+    if (validationsResult.forall(_.isRight)) Right(OmmersValid)
+    else {
+      validationsResult.filterNot(_.isRight) match {
+        case Left(headerError) :: _ => Left(OmmersHeaderError(Some(headerError)))
```
Comment (ten15bit) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: This will never match because `validationsResult.filterNot(_.isRight)` returns a `Seq`.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 9]
```diff
@@ -85,26 +91,28 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
-    else Left(OmmersAncestorsError)
+    lazy val ancestorsParents: Seq[ByteString] = ancestors.map(_.parentHash)
+    lazy val ommersParentsHashes: Seq[ByteString] = ommers.map(_.parentHash)
+
+    // parent not an ancestor or is too old (we only compare up to 6 previous ancestors)
+    lazy val ommersParentsAreAllAncestors: Boolean = ommersParentsHashes.forall(ancestorsParents.contains)
+
+    if (ommersThatAreAncestors.nonEmpty) Left(OmmerIsAncestorError)
+    else if (!ommersParentsAreAllAncestors) Left(OmmerParentIsNotAncestorError)
+    else Right(OmmersValid)
```
Comment (ten15bit) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: I wouldn't default to `Right` in the validation.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I wouldn't remove it...in the worst case if `getNBlocksBack` returns an empty list, the ommers validation will return `Right`.

The `validateOmmersAncestors` method used to return a `OmmersAncestorsError` to indicate there was a problem getting the ancestors...that behavior seems to be lost as well.

File: src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala (hunk 5 above)
Comment (ten15bit) [inline src/main/scala/io/iohk/ethereum/consensus/ethash/validators/OmmersValidator.scala]: This will probably have to contain a list of header errors.

see https://github.com/input-output-hk/mantis/pull/948/files#diff-4ffb3286be2984d1679ee76bbb2e5e404f64147f1d77ab0eb95f8031a1379635R81

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala [hunk 10]
```diff
@@ -414,13 +423,16 @@ class BlockFetcher(
 
   private def handleRequestResult(fallback: FetchMsg)(msg: Any): Task[Any] = msg match {
     case failed: RequestFailed =>
-      log.debug("Request failed due to {}", failed)
+      log.error("Request failed due to {}", failed)
```
Comment (ten15bit) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: `error` is usually used for fatal errors...I think a `warn` would be enough here

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala (hunk 2 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: I agree that not all of them should be changed to `info`...I'd change the ones indicating failures to `warn` and only important ones to info.

**Review by ten15bit — COMMENTED** [review]
Looks good! Some comments regarding the refactoring and log levels.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 8 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: Yes, and I have deconstructed that in a head and tail, and there's a test proving it works https://github.com/input-output-hk/mantis/blob/feature/ETCM-709-improve-ommers-validation/src/test/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidatorSpec.scala#L31-L39
There's never more than one error reported here

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala (hunk 2 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: I have rolled back the modification in the logs, expect one. I think this can be reviewed better by someone that actually touches the code

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: It's because no error was found, what would you return there then?

File: src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala (hunk 10 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockFetcher.scala]: Updated

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: `validateOmmersAncestors` was returning `OmmersUsedBeforeError`, but since two different validations were happening in that method I made it clear by replacing `OmmersUsedBeforeError` by two others that are more precise.

Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: `validateOmmersAncestors` was returning `OmmersAncestorsError` in case `getNBlocksBack` returned an invalid number of results: https://github.com/input-output-hk/mantis/pull/948/files#diff-4ffb3286be2984d1679ee76bbb2e5e404f64147f1d77ab0eb95f8031a1379635L107

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: I gave an example below why defaulting to `Right` is problematic.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 8 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: > here's never more than one error reported here

It's calling `blockHeaderValidator.validate` for every ommer - so why can't there be more than one error?

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: Ah, ok, now I understand your point. The `getNBlocksBack` is a method that is written by Mantis, so I don't understand why we would have to validate it every single time it's called.

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 8 above)
Comment (LeonorLunatech) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:81]: For a couple of reason, the main one was to keep the method signature and the second is because as soon an error is found the block validation is stopped, in general, and I saw no reason to change that

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: Indeed, you are right, it can return more than one error and I can group them all in a list.
I think I was following the approach of stopping the validation at the first error found. I will adjust it to be a list, meaning `case class OmmersHeaderError(error: List[BlockHeaderError])`

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: Ideally `getNBlocksBack` would return a type that allows encoding errors but since it doesn't, it's a good idea to check whether it returns the expected number of blocks.

In theory we should always be able to get the expected numbers of blocks, so if the number doesn't match something is definitely broken. By removing the check and defaulting to `Right` that problem will be silently ignored.

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: So I am adding a check in `private def validateOmmersNotUsed`,making sure that if the list is empty the validation fails, in `private[std] def validateOmmersAncestors` the validation will already fail if the list of ancestors is empty

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I have also added a test named `report failure if getNBlocksBack returns an empty list of ancestors`, involving `validateOmmersAncestors` method

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: Sorry, but where?

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I had to remove the explicit check that checked that the list of ommers from ancestors was not empty, because it was causing 2 ETS tests to fails

**Comment by leo-bogastry** [comment]
Tested FastSync and RegularSync once again in `etc`

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: Sorry, missed the comment. I was referring to https://github.com/input-output-hk/mantis/pull/948#discussion_r609043993

In general I'd say the problem with defaulting to `Right` is that it will also return Right for every unhandled/unexpected case while the other way around is more restrictive, making the errors more visible.

In this specific case I'd re-add at the very least the check that was removed in https://github.com/input-output-hk/mantis/pull/948/files#diff-4ffb3286be2984d1679ee76bbb2e5e404f64147f1d77ab0eb95f8031a1379635L160

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala [hunk 11]
```diff
@@ -66,16 +66,24 @@ class StdOmmersValidator(blockchainConfig: BlockchainConfig, blockHeaderValidato
...
-    if (ommers.forall(blockHeaderValidator.validate(_, getBlockHeaderByHash).isRight)) Right(OmmersValid)
-    else Left(OmmersNotValidError)
+    val validationsResult: Seq[Either[BlockHeaderError, BlockHeaderValid]] =
+      ommers.map(blockHeaderValidator.validate(_, getBlockParentsHeaderByHash))
+
+    if (validationsResult.forall(_.isRight)) Right(OmmersValid)
+    else {
+      val ommersHeaderError = validationsResult
+        .collect { case Left(error) => error }
+        .map(error => OmmersHeaderError(List(error)))
```
Comment (ten15bit) [inline src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: `.map` and `.reduce` are not needed...can be simplified to
```
val errors = validationsResult.collect { case Left(error) => error }.toList
Left(OmmersHeaderError(errors))
```

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 1 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: We shouldn't ignore the issue if blocks aren't found. I'd suggest one of the two approaches

- add a check for the correct size of the result and return an appropriate error otherwise
- merge as it is but create a follow-up ticket to refactor `GetNBlocksBack` to return an error in case blocks aren't found (better approach imho)

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:166]: I created a follow-up ticket ETCM-776

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: That is the same check that will be added in  ETCM-776

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 11 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: You are right, I really overcomplicated that :)

Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala]: :)

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (ten15bit) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: ETCM-776 was created after I wrote that comment.

...and I still think that defaulting to Right in validations is less secure.

**Review by ten15bit — APPROVED** [review]

File: src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala (hunk 9 above)
Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: I understand the philosophy of what you are saying, but I also think that in this case it would really overcomplicate the code. The whole reason why I was not able to understand why a block validation failed in testnet was because those two separate errors `OmmerIsAncestorError` and `OmmerParentIsNotAncestorError` did not exist.
This code is only making these two specific validations otherwise there is no error.

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: What I can propose is a bigger refactoring, than ETCM-776. `GetNBlocksBack` should be part of the `for` comprehension, inside the more generic `validate` method and `validateOmmersAncestors` could be divided into two method (each one returning only one error)

Comment (leo-bogastry) [reply src/main/scala/io/iohk/ethereum/consensus/validators/std/StdOmmersValidator.scala:115]: I have rewritten ETCM-776
